File: dart/travel_angels/lib/models/trending_destination.dart
import 'package:json_annotation/json_annotation.dart';
import 'package:travel_angels/models/document_location.dart';

part 'trending_destination.g.dart';

@JsonSerializable()
class TrendingDestination {
  @JsonKey(name: 'id', fromJson: _idFromJson)
  final String id;

  @JsonKey(name: 'name')
  final String name;

  @JsonKey(name: 'formattedAddress')
  final String formattedAddress;

  @JsonKey(name: 'documentCount')
  final int documentCount;

  @JsonKey(name: 'placeId')
  final String placeId;

  @JsonKey(name: 'latitude')
  final double latitude;

  @JsonKey(name: 'longitude')
  final double longitude;

  @JsonKey(name: 'locationType', fromJson: _locationTypeFromJson)
  final LocationType locationType;

  @JsonKey(name: 'rank')
  final int rank;

  TrendingDestination({
    required this.id,
    required this.name,
    required this.formattedAddress,
    required this.documentCount,
    required this.placeId,
    required this.latitude,
    required this.longitude,
    required this.locationType,
    required this.rank,
  });

  factory TrendingDestination.fromJson(Map<String, dynamic> json) =>
      _$TrendingDestinationFromJson(json);

  Map<String, dynamic> toJson() => _$TrendingDestinationToJson(this);

  static String _idFromJson(dynamic json) {
    if (json == null) return '';
    return json.toString();
  }

  static LocationType _locationTypeFromJson(dynamic json) {
    if (json == null) return LocationType.city;
    final str = json.toString().toLowerCase();
    switch (str) {
      case 'country':
        return LocationType.country;
      case 'continent':
        return LocationType.continent;
      case 'city':
      default:
        return LocationType.city;
    }
  }
}

File: go/pkg/models/trending_destination.go
package models

import (
	"time"

	"github.com/google/uuid"
)

type TrendingDestination struct {
	ID               uuid.UUID    `gorm:"type:uuid;primary_key;" json:"id"`
	CreatedAt        time.Time    `json:"createdAt"`
	UpdatedAt        time.Time    `json:"updatedAt"`
	LocationType     LocationType `gorm:"type:varchar;not null" json:"locationType"`
	PlaceID          string       `gorm:"type:varchar;not null" json:"placeId"`
	Name             string       `gorm:"type:varchar;not null" json:"name"`
	FormattedAddress string       `gorm:"type:text;not null" json:"formattedAddress"`
	DocumentCount    int          `gorm:"type:integer;not null" json:"documentCount"`
	Rank             int          `gorm:"type:integer;not null" json:"rank"`
	Latitude         float64      `gorm:"type:double precision;not null" json:"latitude"`
	Longitude        float64      `gorm:"type:double precision;not null" json:"longitude"`
}

func (TrendingDestination) TableName() string {
	return "trending_destinations"
}
